lib/context.js

```diff
@@ -0,0 +1,270 @@
+'use strict';
+
+const debug = require('debug')('koa-session:context');
+const Session = require('./session');
+const uid = require('uid-safe');
+const util = require('./util');
+
+const ONE_DAY = 24 * 60 * 60 * 1000;
+
+class ContextSession {
+  /**
+   * context session constructor
+   * @api public
+   */
+
+  constructor(ctx, opts) {
+    this.ctx = ctx;
+    this.opts = Object.assign({}, opts);
+    this.store = this.opts.store;
+  }
+
+  /**
+   * internal logic of `ctx.session`
+   * @return {Session} session object
+   *
+   * @api public
+   */
+
+  get() {
+    const session = this.session;
+    // already retrieved
+    if (session) return session;
+    // unset
+    if (session === false) return null;
+
+    // cookie session store
+    if (!this.store) this.initFromCookie();
+    return this.session;
+  }
+
+  /**
+   * internal logic of `ctx.session=`
+   * @param {Object} val session object
+   *
+   * @api public
+   */
+
+  set(val) {
+    if (val === null) {
+      this.session = false;
+      return;
+    }
+    if (typeof val === 'object') {
+      // use the original `externalKey` if exists to avoid waste storage
+      this.create(val, this.externalKey);
+      return;
+    }
+    throw new Error('this.session can only be set as null or an object.');
+  }
+
+  /**
+   * init session from external store
+   * will be called in the front of session middleware
+   *
+   * @api public
+   */
+
+  * initFromExternal() {
+    debug('init from external');
+    const ctx = this.ctx;
+    const opts = this.opts;
+
+    const externalKey = ctx.cookies.get(opts.key, opts);
+    debug('get external key from cookie %s', externalKey);
+
+    if (!externalKey) {
+      // create a new `externalKey`
+      this.create();
+      return;
+    }
+
+    const json = yield this.store.get(externalKey, ctx);
```

Comment:
it could throw here, it is strong dependent on external store if `options.store` present.

I'll remove `ctx` here, it is useless. Keep the signature the same as `koa-generic-session`'s store.